### PR TITLE
edit error handling in create-db(), edit-user()

### PR DESCRIPTION
**Pull request by cmoussa1**
**Problem**: Noted in #38, there was no sort of error handling if `create-db()` failed for whatever reason while creating the database. There was also an instance in `edit-user()` where if somebody tried to edit a user's account information and specified a non-existent field, the program would call `sys.exit(1)`.

---

This PR wraps `create-db()` in a `try-except` block, and will print any SQLite operational error to the screen. It also changes `edit-user()` to raise an `Exception` instead of calling `sys.exit()` if a non-existent field is specified.

I also noticed that `accounting_cli_functions.py` had a couple of unused `import` statements, so I figured I would remove those as well in this PR (if anything, I can always open another tiny PR that removes it instead of including it in this one).

I believe now there is only one instance where `sys.exit()` is called - in `accounting_cli.py`:

```python
# try to open database file; will exit with -1 if database file not found
path = args.path if args.path else "FluxAccounting.db"
try:
    conn = sqlite3.connect("file:" + path + "?mode=rw", uri=True)
except sqlite3.OperationalError:
    print("Unable to open database file")
    sys.exit(-1)
```

I figured not being able to connect to the database file would warrant a `sys.exit()` since essentially none of the subcommands would work if you can't connect to the database file... is this the right intuition to have? Should an `Exception` instead be raised?

### CODE REVIEW

File: accounting/create_db.py
Comment (chu11) [inline accounting/create_db.py:62]: if an exception happens before the `conn.close()` this won't be cleaned up properly.  perhaps should use a `finally` clause here?

File: accounting/create_db.py
Comment (chu11) [inline accounting/create_db.py:64]: IMO you shouldn't `print(e)`.  Raise the exception and let the caller decide if they want to print or not.

**Review by chu11 — COMMENTED** [review]
> I figured not being able to connect to the database file would warrant a sys.exit() since essentially none of the subcommands would work if you can't connect to the database file... is this the right intuition to have? Should an Exception instead be raised?

I think this is fine.

File: accounting/create_db.py
Comment (cmoussa1) [reply accounting/create_db.py]: Oh yeah, I forgot about that. Just pushed a fix that adds a `finally` clause that closes the SQLite connection.

**Comment by cmoussa1** [comment]
Thanks @chu11! I just pushed an update with your suggestions. Thanks for providing your input on the use of `sys.exit()`.

**Comment by chu11** [comment]
Do you have to try-catch the call to `create_db`?

```
    if args.func == "create_db":
        c.create_db(args.dbpath)
        sys.exit(0)

```

**Comment by cmoussa1** [comment]
Hm. I'm not sure. 

Well the `create_db()` function that gets called is itself wrapped in a `try-except`. Does the call to `create_db()` also have to be wrapped in `try-except`?

**Comment by chu11** [comment]
> Well, the create_db() function that gets called is itself wrapped in a try-except. Does the call to create_db() also have to be wrapped in try-except?

Before you `print(e)`'-ed in `create_db()` so I thought you'd want to capture and print out the `e` now.  But you don't necessarily have to.

**Comment by cmoussa1** [comment]
Yeah, `edit-user()`  does the same thing, it just raises the exception, so I figured I might make it consistent with that. Do you think it would be better to instead print out `e`?

**Comment by chu11** [comment]
I'm not sure what is preferred.  Perhaps @SteVwonder has an opinion.  I certainly have met my share of programs that capture exceptions and print something pretty, and others that just happily let the exception & stack trace be output to the user directly.  But unsure if there's a more common approach in the Python world.

**Comment by SteVwonder** [comment]
> I'm not sure what is preferred. Perhaps @SteVwonder has an opinion. I certainly have met my share of programs that capture exceptions and print something pretty, and others that just happily let the exception & stack trace be output to the user directly.

Personally, I don't have a strong preference one way or the other.  Over in flux-core at least, I think we have tried to keep the default to be "pretty" and then provide env vars that turn up the verbosity.  With these tools being more sys admin facing, I could see an argument for being more verbose by default.

FWIW, if you decorate your top-level function (i.e., main) with the `CLIMain` decorator, then the exception handling gets handled automatically for you.  Any uncaught exceptions are "pretty printed" (no stack trace) by default, and the full stack trace is included if `FLUX_PYCLI_LOGLEVEL` is set.  The decorator also sets up the `logger` for you if you want to use that rather than printing directly.  Sorry if you are already aware of all of that and using it.

**Comment by cmoussa1** [comment]
> Personally, I don't have a strong preference one way or the other. Over in flux-core at least, I think we have tried to keep the default to be "pretty" and then provide env vars that turn up the verbosity. With these tools being more sys admin facing, I could see an argument for being more verbose by default.

That makes sense. To me, it seems reasonable to pretty print an error in `edit_user()`'s case (I went ahead and changed it to raise a `ValueError` and print that to the screen). 

In `create_db()`'s case, I wonder if it's reasonable to use `sys.exit()` if the database file can't be created:

```python
try:
    # open connection to database
    logging.info("Creating Flux Accounting DB")
    conn = sqlite3.connect("file:" + filepath + "?mode:rwc", uri=True)
    logging.info("Created Flux Accounting DB successfully")
except sqlite3.OperationalError as e:
    print("Error creating database:", e)
    logging.info(e)
    sys.exit(-1)
```

If there is an error trying to create the database here (for example, if the user doesn't have permission to the directory they are trying to place the database file in), the error will be pretty printed to the screen and written to the log file, but the program will exit before it moves on and tries to create the two tables. I think this will also prevent the `UnboundLocalError` exception that would get raised when it tries to close the connection in the `finally` block. 

Thoughts on this? Would this be an appropriate use case of `sys.exit()`?

**Comment by SteVwonder** [comment]
> In create_db()'s case, I wonder if it's reasonable to use sys.exit() if the database file can't be created:

I think the code snippet you have looks fine.  The print immediately followed by a log seems weird to me.  If you change `logging.info` to `logging.error`, does it print out to the console?  If so, I'd drop the print.  If you need to keep the print, I'd make sure to print to stderr `print("error message", file=sys.stderr)`.

**Comment by cmoussa1** [comment]
> If you change logging.info to logging.error, does it print out to the console?

I don't believe it does, it just gets written to the `.log` file with level `ERROR`:

```
[fluxuser@8e373ae13a64 flux-ba]$ cat db_creation.log 
INFO:root:Creating Flux Accounting DB
ERROR:root:unable to open database file
```

But I don't necessarily need to print anything to the screen, so I can drop the `print()`.

**Comment by cmoussa1** [comment]
Rebased to catch up after #55.

**Comment by dongahn** [comment]
@cmoussa1 got reviews from two people already. Is this good to go? 

I would like to start my current fair-tree effort from a clean slate as much as possible.

**Comment by cmoussa1** [comment]
I just rebased to catch up after #57; this should be good to go on my end. 👍

**Comment by cmoussa1** [comment]
@dongahn pointed out that the output format of the `Exception`s that get raised from `print_hierarchy()` could be improved, so I attached it here (fc7319a) since it is similar to the other fixes attached in this PR.

File: accounting/create_db.py
Comment (chu11) [inline accounting/create_db.py:30]: is there a reason for exiting -1 instead of 1?  is it a python thing?

File: accounting/print_hierarchy.py
Comment (chu11) [inline accounting/print_hierarchy.py:32]: i'm a little confused by this design pattern.  why capture the exception here.  why not capture the exception when `print_full_hierarchy()` is called?

File: accounting/create_db.py
Comment (cmoussa1) [reply accounting/create_db.py:30]: Yeah, I also exited -1 here in **accounting_cli.py**:

```python
# try to open database file; will exit with -1 if database file not found
path = args.path if args.path else "FluxAccounting.db"
try:
    conn = sqlite3.connect("file:" + path + "?mode=rw", uri=True)
except sqlite3.OperationalError:
    print("Unable to open database file")
    sys.exit(-1)
```

File: accounting/print_hierarchy.py
Comment (cmoussa1) [reply accounting/print_hierarchy.py:32]: Are you thinking to instead capture the exception in **accounting_cli.py**?

```python
elif args.func == "print_hierarchy":
    print(ph.print_full_hierarchy(conn))
```

I guess my reasoning was I was capturing exceptions from inside the subcommands themselves, not the program (**accounting_cli.py**) that calls those functions, so I tried to make it consistent. 

Does it make more sense to capture the exception in **accounting_cli.py**?

File: accounting/create_db.py
Comment (chu11) [reply accounting/create_db.py:30]: i would say just do `sys.exit(1)` for errors.  I think it's the standard convention unless there's a special python convention I'm unaware of: https://tldp.org/LDP/abs/html/exitcodes.html

File: accounting/print_hierarchy.py
Comment (chu11) [reply accounting/print_hierarchy.py:32]: ok.  In that case, perhaps I'd suggest;

```
try:
  if ...:
     raise ...
  elif ...:
     raise ...
except:
   print(e)
   sys.exit(1)
```

so no need for the double try/exception.  I think that double try/except is what confused me the most.

Comment (cmoussa1) [reply accounting/print_hierarchy.py]: Ah, okay. Thanks. Sorry for the confusion. That change should be pushed up now.

**Comment by cmoussa1** [comment]
pushed a small change that adds a couple of unit tests that interact with the `Exception`s that get thrown in `print_full_hierarchy()`.

**Review by chu11 — APPROVED** [review]
LGTM, set MWP when reading

**Pull request merged**
